platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+                    ( const void * ) pMbedtlsPkcs11Credentials ) );
+        returnStatus = MBEDTLS_PKCS11_INVALID_PARAMETER;
+    }
+    else
+    {
+        snprintf( portStr, 6, "%u", port );
+        pMbedtlsPkcs11Context = pNetworkContext->pParams;
+
+        /* Configure MbedTLS. */
+        returnStatus = configureMbedtls( pMbedtlsPkcs11Context, pHostName, pMbedtlsPkcs11Credentials, recvTimeoutMs );
+    }
+
+    /* Establish a TCP connection with the server. */
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        mbedtlsError = mbedtls_net_connect( &( pMbedtlsPkcs11Context->socketContext ),
+                                            pHostName,
+                                            portStr,
+                                            MBEDTLS_NET_PROTO_TCP );
+
+        if( mbedtlsError != 0 )
+        {
+            LogError( ( "Failed to connect to %s with error %d.", pHostName, mbedtlsError ) );
+            returnStatus = MBEDTLS_PKCS11_CONNECT_FAILURE;
+        }
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        /* Perform the TLS handshake. */
+        do
+        {
+            mbedtlsError = mbedtls_ssl_handshake( &( pMbedtlsPkcs11Context->context ) );
+        } while( ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_READ ) ||
+                 ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_WRITE ) );
+
+        if( ( mbedtlsError != 0 ) || ( mbedtls_ssl_get_verify_result( &( pMbedtlsPkcs11Context->context ) ) != 0U ) )
+        {
+            LogError( ( "Failed to perform TLS handshake: mbedTLSError= %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+            returnStatus = MBEDTLS_PKCS11_HANDSHAKE_FAILED;
+        }
+    }
+
+    /* Clean up on failure. */
+    if( returnStatus != MBEDTLS_PKCS11_SUCCESS )
+    {
+        contextFree( pMbedtlsPkcs11Context );
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) Connection to %s established.", pNetworkContext, pHostName ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void Mbedtls_Pkcs11_Disconnect( NetworkContext_t * pNetworkContext )
+{
+    MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context = NULL;
+    int tlsStatus = 0;
+
+    if( ( pNetworkContext != NULL ) && ( pNetworkContext->pParams != NULL ) )
+    {
+        pMbedtlsPkcs11Context = pNetworkContext->pParams;
+        /* Attempting to terminate TLS connection. */
+        tlsStatus = mbedtls_ssl_close_notify( &( pMbedtlsPkcs11Context->context ) );
+
+        /* Ignore the WANT_READ and WANT_WRITE return values. */
+        if( ( tlsStatus != MBEDTLS_ERR_SSL_WANT_READ ) &&
+            ( tlsStatus != MBEDTLS_ERR_SSL_WANT_WRITE ) )
+        {
+            if( tlsStatus == 0 )
+            {
+                LogInfo( ( "(Network connection %p) TLS close-notify sent.",
```

Comment:
The logging of the network context pointer in all the logs can be avoided as it does not provide any debug value